### PR TITLE
New voting tutorial

### LINKED ISSUE #376: Update Voting and Kitty Hats Tutorials
### Issue To Be Solved
Update Voting and Kitty Hats Tutorial to:

* Use more sophisticated composability features (run time types)
* Use latest Cadence best practices
* Fix bugs
* Better explanations for features

### PR DESCRIPTION
**Pull request by ggpt-labs**
Closes https://github.com/onflow/flow/issues/376

### CODE REVIEW

File: tutorials/08-voting/contracts/GovernanceToken.cdc [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+// GovernanceToken.cdc
+//
+// The GovernanceToken contract is a sample implementation of a fungible token on Flow which can be used for voting.
+
+pub contract GovernanceToken: FungibleToken {
+    // checkpointCounter is incremented with each proposal. 
+    // For voting, only the user's balance before 
+    // the respective proposal checkpoint value will be taken into account.
+    // TODO: this should not be public
+    pub(set) var checkpointCounter: UInt16
```
Comment (joshuahannan) [inline tutorials/08-voting/contracts/GovernanceToken.cdc:12]: Isn't this supposed to be a timestamp? Do you still want to make this an emulator/CLI tutorial instead of a playground tutorial?

File: tutorials/08-voting/contracts/GovernanceToken.cdc [hunk 2]
```diff
@@ -0,0 +1,200 @@
+import FungibleToken from "./FungibleToken.cdc"
+
+// GovernanceToken.cdc
+//
+// The GovernanceToken contract is a sample implementation of a fungible token on Flow which can be used for voting.
+
+pub contract GovernanceToken: FungibleToken {
```
Comment (joshuahannan) [inline tutorials/08-voting/contracts/GovernanceToken.cdc:7]: I hesitate creating a tutorial that implements the Fungible Token standard, because in the user's learning journey, we haven't actually explained how the fungible token standard works yet, so it might be introducing more complexity than is worth introducing. What do you think?

File: tutorials/08-voting/contracts/GovernanceToken.cdc [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    // Function that creates a new Vault with a balance of zero
+    // and returns it to the calling context. A user must call this function
+    // and store the returned Vault in their storage in order to allow their
+    // account to be able to receive deposits of this token type.
+    //
+    pub fun createEmptyVault(): @FungibleToken.Vault {
+        return <-create Vault(balance: 0.0)
+    }
+
+	// VaultMinter
```
Comment (joshuahannan) [inline tutorials/08-voting/contracts/GovernanceToken.cdc:136]:
```suggestion
// VaultMinter
```

File: tutorials/08-voting/contracts/Voting.cdc [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    // paths
+    pub let adminStoragePath: StoragePath
+    pub let ballotStoragePath: StoragePath
+    pub let ballotPublicPath: PublicPath
+
+    pub struct ProposalData {
+        pub let name: String
+        pub let checkpoint: UInt16
+        pub(set) var votes: UFix64
+        pub(set) var voters: {UInt64: Bool}
```
Comment (joshuahannan) [inline tutorials/08-voting/contracts/Voting.cdc:35]: Can you add comments to show what each line means?

File: tutorials/08-voting/contracts/Voting.cdc [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+*
+*   Every user with a ballot is allowed to approve their chosen proposals.
+*   A user can choose their votes and cast them
+*   with the tx_05_SelectAndCastVotes.cdc transaction.
+*
+*/
+
+import GovernanceToken from "./GovernanceToken.cdc"
+
+pub contract Voting {
```
Comment (joshuahannan) [inline tutorials/08-voting/contracts/Voting.cdc:21]: Can you make this a more specific name? like `CadenceExampleVotingTutorial`

File: tutorials/08-voting/contracts/GovernanceToken.cdc (hunk 2 above)
Comment (joshuahannan) [inline tutorials/08-voting/contracts/GovernanceToken.cdc:7]: We should make this a more specific name, like `CadenceGovernanceTokenTutorial`

File: tutorials/08-voting/contracts/Voting.cdc [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    // This is the resource that is issued to users.
+    // When a user gets a Ballot resource, they call the `vote` function
+    // to include their votes
+    pub resource Ballot: Votable {
+        // id of GovernanceToken Vault
+        pub let vaultId: UInt64
+        // array of GovernanceToken Vault's votingWeightDataSnapshot
+        pub let votingWeightDataSnapshot: [GovernanceToken.VotingWeightData]
```
Comment (joshuahannan) [inline tutorials/08-voting/contracts/Voting.cdc:66]: Might be better to make this a dictionary keyed by the checkpoint number to make it easier to lookup

File: tutorials/08-voting/contracts/Voting.cdc [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    // initializes the contract by setting the proposals and votes to empty
+    // and creating a new Admin resource to put in storage
+    init() {
+        self.proposals = []
+
+        self.ballotStoragePath = /storage/Ballot
+        self.adminStoragePath = /storage/VotingAdmin
+        self.ballotPublicPath = /public/Ballot
```
Comment (joshuahannan) [inline tutorials/08-voting/contracts/Voting.cdc:125]: I would also make these path values more specific

**Review by joshuahannan — COMMENTED** [review]
Great work so far! I have a few high level design comments and some smaller detail comments.

File: tutorials/08-voting/contracts/GovernanceToken.cdc (hunk 1 above)
Comment (gus-labs) [reply tutorials/08-voting/contracts/GovernanceToken.cdc:12]: Yes! 

I actually did it in the very first draft: https://github.com/onflow/voting-tutorial/blob/main/tutorials/08-voting/contracts/GovernanceToken.cdc#L36

I changed after I was informed about the idea of merging this previous PR code into mine: https://github.com/onflow/flow-playground-tutorials/pull/8/files.

But in reality I would like to use timestamp. I will change to it.

File: tutorials/08-voting/contracts/Voting.cdc (hunk 5 above)
Comment (gus-labs) [reply tutorials/08-voting/contracts/Voting.cdc:21]: So far, in all other tutorials, only path names contained a reference to 'Cadence' and 'Tutorial'. Are you sure we should rename the contracts?

File: tutorials/08-voting/contracts/Voting.cdc (hunk 6 above)
Comment (gus-labs) [reply tutorials/08-voting/contracts/Voting.cdc:66]: Not using a checkpoint anymore but the block timestamp, so unfortunately, dictionary lookup won't work.

File: tutorials/08-voting/contracts/Voting.cdc (hunk 7 above)
Comment (gus-labs) [reply tutorials/08-voting/contracts/Voting.cdc]: Done.

File: tutorials/08-voting/contracts/GovernanceToken.cdc (hunk 2 above)
Comment (gus-labs) [reply tutorials/08-voting/contracts/GovernanceToken.cdc:7]: I got the idea to use the Fungible Token contract because Satyam and you were mentioning it in the issue discussion. 
I don't really have a strong opinion here.

File: tutorials/08-voting/contracts/VotingTutorialAdministration.cdc [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+*   A user can choose their votes and cast them
+*   with the tx_05_SelectAndCastVotes.cdc transaction.
+*
+*/
+
+import VotingTutorialGovernanceToken from "./VotingTutorialGovernanceToken.cdc"
+
+pub contract VotingTutorialAdministration {
+
+    // dictionary of proposals to be approved
```
Comment (satyamakgec) [inline tutorials/08-voting/contracts/VotingTutorialAdministration.cdc:23]: Sorry for nitpicks - It is always beautiful to start with the capital letter as the code comment statement

File: tutorials/08-voting/contracts/VotingTutorialAdministration.cdc [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        // used to record if a voter as represented by the vault id has already voted
+        pub(set) var voters: {UInt64: Bool}
+
+        init(name: String, options: [String], blockTs: UFix64) {
+            self.name = name
+            self.options = options
+            self.blockTs = blockTs
+            self.votes = {}
+            for index, option in options {
+                self.votes[index] = 0.0
```
Comment (satyamakgec) [inline tutorials/08-voting/contracts/VotingTutorialAdministration.cdc:48]: Do we need this?

File: tutorials/08-voting/contracts/VotingTutorialAdministration.cdc [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    pub resource interface Votable {
+        pub vaultId: UInt64
+        pub votingWeightDataSnapshot: [VotingTutorialGovernanceToken.VotingWeightData]
+
+        pub fun vote(proposalId: Int, optionId: Int){
+            pre {
+                VotingTutorialAdministration.proposals[proposalId] != nil: "Cannot vote for a proposal that doesn't exist"
+                VotingTutorialAdministration.proposals[proposalId]!.voters[self.vaultId] == nil: "Cannot cast vote again using same Governance Token Vault"
+                optionId < VotingTutorialAdministration.proposals[proposalId]!.options.length: "This option does not exist"
+                self.votingWeightDataSnapshot != nil && self.votingWeightDataSnapshot.length > 0: "Can only vote if balance exists"
```
Comment (satyamakgec) [inline tutorials/08-voting/contracts/VotingTutorialAdministration.cdc:63]: I think we don't need a `nil` check here.

**Review by satyamakgec — COMMENTED** [review]
Hey @gus-labs can you please add more documentation related to the VotingAdministration contract? It is hard to follow there. I am gonna do the second review once we have that documentation. 

We can also schedule a call related to the design of the smart contracts.

File: tutorials/08-voting/contracts/VotingTutorialAdministration.cdc (hunk 9 above)
Comment (gus-labs) [reply tutorials/08-voting/contracts/VotingTutorialAdministration.cdc:48]: Yes, because the optional value is force unwrapped in line 99: https://github.com/gus-labs/flow-playground-tutorials/blob/18253165fbaaadc29421acc026b9eee3d6291180/tutorials/08-voting/contracts/VotingTutorialAdministration.cdc#L99

File: tutorials/07-voting/README.md [hunk 11]
```diff
@@ -0,0 +1,645 @@
+---
+title: 9. Voting Contract
+---
+
+With the advent of blockchain technology and smart contracts,
+it has become popular to try to create decentralized voting mechanisms that allow large groups of users to vote completely on chain.
+This tutorial will provide a trivial example for how this might be achieved by using a resource-oriented programming model.
+Two contracts will allow users to vote on multiple proposals while their voting power is determined by the balance of certain tokens. 
+These so called governance tokens track the user account balance over time, thus enabling insight into the balance just before the creation of a proposal.
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md:9]: Also, it would be good to mention that this tutorial provides an introduction to the CLI and emulator

File: tutorials/07-voting/README.md [hunk 12]
```diff
@@ -0,0 +1,645 @@
...
+3. Create voter accounts for these users
+4. Mint tokens to these accounts
+5. Create proposals for users to vote on
+6. Create `Ballot` resources for both voters
+7. Record and cast votes in the central voting contract
+8. Read the results of the vote
+
+## Setup the environment
+
+In this tutorial, we are introducing the [Flow CLI](https://developers.flow.com/tools/flow-cli/index), which allows us to run our code on a local blockchain emulator.
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md]: I think we need a little bigger description of the CLI and emulator. For example, explain that the CLI is a easy way to run transactions, scripts, query state from any blockchain instance, including mainnet, testnet, and the emulator. And the emulator runs a simplified local blockchain for testing but still has most of the same interfaces and functionality

File: tutorials/07-voting/README.md [hunk 13]
```diff
@@ -0,0 +1,645 @@
...
+
+In this tutorial, we are introducing the [Flow CLI](https://developers.flow.com/tools/flow-cli/index), which allows us to run our code on a local blockchain emulator. 
+Please follow the link and install it on your computer. You should now be able to call it by simply entering `flow`, which will show you a list with all commands.
+The CLI also contains the blockchain emulator, please take a look at the [Flow CLI subsection](https://developers.flow.com/tools/flow-cli/start-emulator) for a brief overview, more details are covered in the [ReadMe](https://github.com/onflow/flow-emulator/#readme).
+
+Once installed, please download the project in the terminal by executing `git clone git@github.com:onflow/flow-playground-tutorials.git`, and then go to the project folder: `cd tutorials/08-voting`. 
+Once there, please execute this command which will create an initial configuration:
+
+```console
+flow init
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md]: need more context on what `flow init` does

File: tutorials/07-voting/README.md [hunk 14]
```diff
@@ -0,0 +1,645 @@
...
+```
+
+Then, edit the configuration and add both the `contracts` and the `deployments` section, this is needed for the deployment of the project:
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md]: We need a description about what flow.json is used for and what the different fields are

File: tutorials/07-voting/README.md [hunk 15]
```diff
@@ -0,0 +1,645 @@
...
+    }
+}
+```
+
+## Run the emulator
+
+Once the configuration is saved, you can start the emulator:
+
+```console
+flow emulator
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md]: more context on what `flow emulator` does

File: tutorials/07-voting/README.md [hunk 16]
```diff
@@ -0,0 +1,645 @@
...
+```
+
+## Deploy the project
+
+Open another terminal window (check that you remain in the project folder) and deploy the project:
+
+```console
+flow project deploy
+```
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md:81]: need more context on what `flow project deploy` does.

File: tutorials/07-voting/README.md (hunk 11 above)
Comment (joshuahannan) [reply tutorials/07-voting/README.md:9]: And mention here that we are assuming that the user understands how the fungible token standard works and if they don't, they need to go to a link that you provide

File: tutorials/07-voting/README.md [hunk 17]
```diff
@@ -0,0 +1,645 @@
...
+
+All three contracts should be deployed now.
+FungibleToken is a Cadence standard, while `VotingTutorialGovernanceToken` and `VotingTutorialAdministration` are specific to this tutorial.
+`VotingTutorialGovernanceToken` is needed in order to vote, it's balance is determining the weight of the vote.
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md]:
```suggestion
`VotingTutorialGovernanceToken` is needed in order to vote, its balance determines the weight of the vote.
```

File: tutorials/07-voting/README.md [hunk 18]
```diff
@@ -0,0 +1,645 @@
...
+`VotingTutorialAdministration` is used for administration of the whole voting process.
+
+`VotingTutorialGovernanceToken` contains the usual vault functionality of the FungibleToken contract and adds a history of voting weight that is updated on each transfer.
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md]: We need a lot more than just a one sentence description of what the contract does. We need you to explain that most of it is identical to a regular fungible token implmentation, but then also walk through each important part that has been added for governance and explain what each of those is for, with code snippet examples to illustrate it

File: tutorials/07-voting/README.md [hunk 19]
```diff
@@ -0,0 +1,645 @@
...
+        emit TokensInitialized(initialSupply: self.totalSupply)
+    }
+}
+```
+
+`VotingTutorialAdministration` contains a struct `ProposalData` which is used to store both the proposal and the total votes.
+Then we have two resources, a `Ballot` and an `Administrator` resource.
+A user can request a `Ballot`, and then vote for a proposal, effectively tallying the vote weight in this contract.
+Using a resource type is logical for this application, because if a user wants to delegate their vote, they can send that `Ballot` to another account.
+Access to the `Administrator` resource is needed in order to add proposals.
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md]: Same comment applies here that I made about the token contract. We need a whole long section here that goes into detail about each part of the contract and explains each feature with code samples.

File: tutorials/07-voting/README.md [hunk 20]
```diff
@@ -0,0 +1,645 @@
...
+}
+```
+
+## Create two extra accounts ('acct2', 'acct3')
+
+The following command creates a user account, you need to execute it twice, once for account 'acct2', and again for 'acct3'.
+When asked, please choose the option *Local Emulator*.
+
+```console
+flow accounts create
```
Comment (joshuahannan) [inline tutorials/07-voting/README.md]: need more context on what `flow accounts create` does

**Review by joshuahannan — COMMENTED** [review]
Remember, we are assuming here that the users who are reading this tutorial have no previous blockchain knowledge and the only things they know are the things that have already been covered in the previous tutorials. We need a lot more elaboration on most of the commands and new concepts. Provide links to the formal documentation for new CLI commands when they are used also.
Transactions also need more explanation to describe what each part of each transaction is doing. 

I didn't finish reviewing the whole tutorial because the changes I'm describing will change it a lot more. Let's try to be very thorough and thoughtful with how we introduce all these new concepts and contracts to users.

File: tutorials/07-voting/README.md (hunk 18 above)
Comment (ggpt-labs) [reply tutorials/07-voting/README.md]: I now added more explanation and code snippets.

File: tutorials/07-voting/README.md (hunk 19 above)
Comment (ggpt-labs) [reply tutorials/07-voting/README.md]: I have now included one code snippet. More snippets would result pretty quick in pasting the whole contract code I think. Instead, I resorted to describe the various components in more detail.
